File: promptc/library.py
"""Library management for prompt libraries."""

import os
from pathlib import Path
from typing import Optional, List


class LibraryManager:
    """Manages prompt library resolution and loading."""

    def __init__(self, project_dir: Optional[Path] = None):
        """
        Initialize library manager.

        Args:
            project_dir: Project directory to look for local prompts
        """
        self.project_dir = project_dir or Path.cwd()
        self.search_paths = self._get_search_paths()

    def _get_search_paths(self) -> List[Path]:
        """Get ordered list of paths to search for prompt libraries."""
        paths = []

        # 1. Project-local prompts
        project_prompts = self.project_dir / "prompts"
        if project_prompts.exists():
            paths.append(project_prompts)

        # 2. Global user prompts
        home_prompts = Path.home() / ".prompts"
        if home_prompts.exists():
            paths.append(home_prompts)

        # 3. Built-in defaults (in package)
        package_dir = Path(__file__).parent
        default_prompts = package_dir / "defaults"
        if default_prompts.exists():
            paths.append(default_prompts)

        return paths

    def resolve(self, import_name: str) -> Optional[Path]:
        """
        Resolve an import name to a file path.

        Args:
            import_name: Import name like "patterns.rest_api" or "company.standards"

        Returns:
            Path to the prompt file, or None if not found
        """
        # Convert dot notation to path (patterns.rest_api -> patterns/rest_api.prompt)
        relative_path = import_name.replace(".", os.sep) + ".prompt"

        for search_path in self.search_paths:
            candidate = search_path / relative_path
            if candidate.exists():
                return candidate

        return None

    def load(self, import_name: str) -> str:
        """
        Load prompt content from a library.

        Args:
            import_name: Import name to load

        Returns:
            Content of the prompt file

        Raises:
            FileNotFoundError: If the import cannot be resolved
        """
        path = self.resolve(import_name)
        if path is None:
            raise FileNotFoundError(
                f"Cannot resolve import '{import_name}'. "
                f"Searched in: {', '.join(str(p) for p in self.search_paths)}"
            )

        return path.read_text(encoding="utf-8")

    def list_libraries(self) -> dict:
        """
        List all available prompt libraries organized by source.

        Returns:
            Dictionary with sources as keys and lists of libraries as values
        """
        libraries = {
            "project": [],
            "global": [],
            "built-in": []
        }

        for search_path in self.search_paths:
            source_key = self._get_source_key(search_path)
            libs = self._scan_directory(search_path)
            libraries[source_key].extend(libs)

        return libraries

    def _get_source_key(self, path: Path) -> str:
        """Get the source key for a search path."""
        if "defaults" in str(path):
            return "built-in"
        elif path.name == "prompts" and path.parent == self.project_dir:
            return "project"
        else:
            return "global"

    def _scan_directory(self, base_path: Path, prefix: str = "") -> List[str]:
        """
        Recursively scan directory for .prompt files.

        Args:
            base_path: Base directory to scan
            prefix: Prefix for library names (for nested directories)

        Returns:
            List of library import names
        """
        libraries = []

        if not base_path.exists():
            return libraries

        for item in sorted(base_path.iterdir()):
            if item.is_file() and item.suffix == ".prompt":
                # Convert to import name
                lib_name = item.stem
                if prefix:
                    lib_name = f"{prefix}.{lib_name}"
                libraries.append(lib_name)
            elif item.is_dir() and not item.name.startswith("."):
                # Recursively scan subdirectories
                subdir_prefix = item.name if not prefix else f"{prefix}.{item.name}"
                libraries.extend(self._scan_directory(item, subdir_prefix))

        return libraries

File: internal/library/library.go
package library

import (
	"embed"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

//go:embed defaults/**/*.prompt
var defaultLibraries embed.FS

// Manager handles prompt library resolution and loading
type Manager struct {
	ProjectDir  string
	SearchPaths []string
}

// NewManager creates a new library manager
func NewManager(projectDir string) *Manager {
	if projectDir == "" {
		var err error
		projectDir, err = os.Getwd()
		if err != nil {
			projectDir = "."
		}
	}

	m := &Manager{
		ProjectDir: projectDir,
	}
	m.SearchPaths = m.getSearchPaths()
	return m
}

func (m *Manager) getSearchPaths() []string {
	var paths []string

	// 1. Project-local prompts
	projectPrompts := filepath.Join(m.ProjectDir, "prompts")
	if _, err := os.Stat(projectPrompts); err == nil {
		paths = append(paths, projectPrompts)
	}

	// 2. Global user prompts
	homeDir, err := os.UserHomeDir()
	if err == nil {
		homePrompts := filepath.Join(homeDir, ".prompts")
		if _, err := os.Stat(homePrompts); err == nil {
			paths = append(paths, homePrompts)
		}
	}

	// 3. Built-in defaults (embedded)
	paths = append(paths, "defaults")

	return paths
}

// Resolve resolves an import name to file content
func (m *Manager) Resolve(importName string) (string, error) {
	// Convert dot notation to path (patterns.rest_api -> patterns/rest_api.prompt)
	relativePath := strings.ReplaceAll(importName, ".", string(filepath.Separator)) + ".prompt"

	// Try each search path
	for _, searchPath := range m.SearchPaths {
		if searchPath == "defaults" {
			// Try embedded files
			content, err := m.loadEmbedded(relativePath)
			if err == nil {
				return content, nil
			}
		} else {
			// Try filesystem
			fullPath := filepath.Join(searchPath, relativePath)
			content, err := os.ReadFile(fullPath)
			if err == nil {
				return string(content), nil
			}
		}
	}

	return "", fmt.Errorf("cannot resolve import '%s'. Searched in: %s",
		importName, strings.Join(m.SearchPaths, ", "))
}

func (m *Manager) loadEmbedded(relativePath string) (string, error) {
	path := filepath.Join("defaults", relativePath)
	content, err := defaultLibraries.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

// Library represents available libraries by source
type Library struct {
	Project  []string
	Global   []string
	BuiltIn  []string
}

// ListLibraries returns all available libraries organized by source
func (m *Manager) ListLibraries() *Library {
	lib := &Library{
		Project:  []string{},
		Global:   []string{},
		BuiltIn:  []string{},
	}

	for _, searchPath := range m.SearchPaths {
		if searchPath == "defaults" {
			lib.BuiltIn = m.scanEmbedded()
		} else {
			libs := m.scanDirectory(searchPath, "")
			if strings.Contains(searchPath, "prompts") && filepath.Dir(searchPath) == m.ProjectDir {
				lib.Project = libs
			} else {
				lib.Global = libs
			}
		}
	}

	return lib
}

func (m *Manager) scanDirectory(basePath, prefix string) []string {
	var libraries []string

	entries, err := os.ReadDir(basePath)
	if err != nil {
		return libraries
	}

	for _, entry := range entries {
		name := entry.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}

		if entry.IsDir() {
			subPrefix := name
			if prefix != "" {
				subPrefix = prefix + "." + name
			}
			libraries = append(libraries, m.scanDirectory(filepath.Join(basePath, name), subPrefix)...)
		} else if strings.HasSuffix(name, ".prompt") {
			libName := strings.TrimSuffix(name, ".prompt")
			if prefix != "" {
				libName = prefix + "." + libName
			}
			libraries = append(libraries, libName)
		}
	}

	sort.Strings(libraries)
	return libraries
}

func (m *Manager) scanEmbedded() []string {
	var libraries []string

	entries, err := defaultLibraries.ReadDir("defaults")
	if err != nil {
		return libraries
	}

	for _, entry := range entries {
		if entry.IsDir() {
			libraries = append(libraries, m.scanEmbeddedDir("defaults/"+entry.Name(), entry.Name())...)
		}
	}

	sort.Strings(libraries)
	return libraries
}

func (m *Manager) scanEmbeddedDir(path, prefix string) []string {
	var libraries []string

	entries, err := defaultLibraries.ReadDir(path)
	if err != nil {
		return libraries
	}

	for _, entry := range entries {
		name := entry.Name()
		if entry.IsDir() {
			subPrefix := prefix + "." + name
			libraries = append(libraries, m.scanEmbeddedDir(path+"/"+name, subPrefix)...)
		} else if strings.HasSuffix(name, ".prompt") {
			libName := prefix + "." + strings.TrimSuffix(name, ".prompt")
			libraries = append(libraries, libName)
		}
	}

	return libraries
}
